nutils/evaluable.py

```diff
@@ -3236,6 +3236,40 @@ def _sum(self, axis):
   def _assparse(self):
     return tuple((*indices[:-1], *divmod(indices[-1], appendaxes(self.shape[-1], values.shape)), values) for *indices, values in self.func._assparse)
 
+class RavelIndex(Array):
+
+  @types.apply_annotations
+  def __init__(self, ia:asarray, ib:asarray, na:asindex, nb:asindex):
+    self._ia = ia
+    self._ib = ib
+    self._na = na
+    self._nb = nb
+    self._length = na * nb
+    super().__init__(args=[ia, ib, nb], shape=ia.shape + ib.shape, dtype=int)
+
+  def evalf(self, ia, ib, nb):
+    return ia[(...,)+(numpy.newaxis,)*ib.ndim] * nb + ib
+
+  def _take(self, index, axis):
+    if axis < self._ia.ndim:
+      return RavelIndex(_take(self._ia, index, axis), self._ib, self._na, self._nb)
+    else:
+      return RavelIndex(self._ia, _take(self._ib, index, axis - self._ia.ndim), self._na, self._nb)
+
+  def _rtake(self, func, axis):
+    if equalindex(func.shape[-1], self._length):
+      return Take(_take(Unravel(func, self._na, self._nb), self._ia, -2), self._ib)
+
+  def _rinflate(self, func, length, axis):
+    if equalindex(length, self._length):
+      return Ravel(Inflate(_inflate(func, self._ia, self._na, func.ndim - self.ndim), self._ib, self._nb))
+
+  def _intbounds_impl(self):
+    nbmin, nbmax = self._nb._intbounds
+    iamin, iamax = self._ia._intbounds
+    ibmin, ibmax = self._ib._intbounds
+    return iamin * nbmin + ibmin, iamax * nbmax + ibmax
```

Comment:
If `iamax` is zero and `nbmax` is inf, then the upper bound is NaN. In this case the upper bound should be `ibmax`.

Reply:
fixed